### DISCUSSION TITLE
Getting bus error when inserting a lot of images

### DISCUSSION THREAD

**Discussion by antoniogamizdelgado** [libvips/libvips#2444] [Q&A]
Hi, I'm a quite new user to this library, but until this point is amazing! Thanks a lot for your hard work. Nonetheless, I'm having some trouble inserting a high number of images into another one. I suppose this is because I'm not using the library as I should, so I'm going to show you my code:

~~~c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <vips/vips.h>


int randomTileCoordinate() { return rand() % 257; }

void drawRandomMarker(VipsImage *main, VipsImage *marker, VipsImage **out) {
    int x = randomTileCoordinate();
    int y = randomTileCoordinate();
    if (vips_insert(main, marker, out, x, y, NULL))
        vips_error_exit(NULL);
}

int
main(int argc, char **argv) {
    srand(time(NULL));

    char *output = argv[1];
    char *marker_path = argv[2];
    char *blank_png_path = argv[3];

    VipsImage *blank, *marker, *out;

    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    if (!(blank = vips_image_new_from_file(blank_png_path, NULL)))
        vips_error_exit(NULL);

    if (!(marker = vips_image_new_from_file(marker_path, NULL)))
        vips_error_exit(NULL);

    drawRandomMarker(blank, marker, &out);

    for (int i = 0; i < 300; i++)
        drawRandomMarker(out, marker, &out);

    if (vips_image_write_to_file(out, output, NULL))
        vips_error_exit(NULL);

    g_object_unref(out);
    g_object_unref(marker);
    g_object_unref(blank);

    return (0);
}
~~~

I'm just basically taking a blank.png of 256x256 pixels, and I'm inserting a marker on certain random positions. This works just fine when I put together 1500:

![image](https://user-images.githubusercontent.com/73933988/133930250-b991a741-3395-464e-95d7-5a00fd1738cf.png)

But it starts to throw a a `bus error` when I try from 1500, like 1600, for instance. I suppose it's a stack overflow or something similar, but I'm not really sure. 

I have also a problem whith the background. The marker is a SVG with a transparent background, but when two markers overlap, the background transparent of one marker hides the other.  I have seen the optional option `background`, But I'm having some troubles finding examples about its use in C.

So, what am I doing wrong? Is `insert` the best operation in this case? Thanks in advance in you answer this question and keep doing a great job improving this amazing library!!! :DDD

**Comment by jcupitt** [comment]
Hello @antoniogamizbadger,

You're stringing 1500 `insert` operations together one after the other. When libvips evaluates your pipeline, it's having to recurse very deeply, and that's triggering a stack overflow.

I would use `composite` instead --- it takes an array of images to insert as an argument, so you can compose very large numbers of images in just one call. For example:

```python
#!/usr/bin/python3

import sys
import random
import pyvips

if len(sys.argv) != 5:
    print(f"usage: {sys.argv[0]}: output-file input-file overlay number")

image = pyvips.Image.new_from_file(sys.argv[2], access="sequential")
overlay = pyvips.Image.new_from_file(sys.argv[3])

n_overlay = int(sys.argv[4])
image = image.composite([overlay] * n_overlay, "over",
    x=[random.randint(0, image.width) for _ in range(n_overlay)],
    y=[random.randint(0, image.height) for _ in range(n_overlay)])

image.write_to_file(sys.argv[1])
```

I can run:

```
$ time ./pin.py x.png ~/pics/Gugg_coloured.jpg ~/pics/Google_Maps_pin.svg 2000
real	0m4.674s
user	0m5.235s
sys	0m0.181s
```

To make:

![x](https://user-images.githubusercontent.com/580843/133935561-80462a54-93a7-4acf-9096-0680963f2099.png)
